ocf_datapipes/transform/xarray/pv/assign_daynight_status.py

```diff
@@ -0,0 +1,84 @@
+import logging
+from datetime import datetime
+from typing import Optional
+
+import numpy as np
+import xarray as xr
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import IterDataPipe
+
+logger = logging.getLogger(__name__)
+
+uk_daynight_dict = {
+    1: [7, 17],
+    2: [7, 18],
+    3: [9, 19],
+    4: [6, 20],
+    5: [5, 21],
+    6: [4, 22],
+    7: [5, 22],
+    8: [5, 21],
+    9: [6, 20],
+    10: [7, 21],
+    11: [7, 17],
+    12: [7, 17],
+}
+logger.info(
+    f"\nThe day and night standard hours are set by https://www.timeanddate.com/sun/uk/london, {uk_daynight_dict}\n"
+)
+
+
+@functional_datapipe("assign_daynight_status")
+class AssignDayNightStatusIterDataPipe(IterDataPipe):
+    """Adds a new dimension of day/night status"""
+
+    def __init__(self, source_datapipe: IterDataPipe):
+        """
+        This method adds extra coordinate of day night status.
+
+        Args:
+            source: Datapipe emiiting Xarray Dataset
+
+        Result:
+            <xarray.Dataset>
+            Dimensions:   (datetime: 289)
+            Coordinates:
+            * datetime  (datetime) datetime64[ns]
+            * status_day (datetime)
+        """
+
+        self.source_datapipe = source_datapipe
+
+    def __iter__(self) -> xr.DataArray():
+        """Returns an xarray dataset with extra dimesion"""
+        logger.info(f"\nReading the Xarray dataset\n")
+        for xr_dataset in self.source_datapipe:
+
+            logger.info(f"\nGetting all the 'time_utc' datetime coordinates\n")
+
+            dates = xr_dataset.coords["time_utc"].values
+
+            logger.info(f"\n Getting Month and Hour values from 'time_utc'and stacking them\n")
+
+            date_month = np.asarray(xr_dataset.time_utc.dt.month.values, dtype=int)
+            date_hr = np.asarray(xr_dataset.time_utc.dt.hour.values, dtype=int)
+            month_hr_stack = np.stack((date_month, date_hr))
+
+            logger.info(f"\nGetting the status of day/night for each timestamp in the timeseries\n")
+
+            status_daynight = []
+            for i in range(len(dates)):
```

Comment:
My biggest concern about this loop is for a year of data, this will take a long time. 

We should be able to do it in the xarray world, or atleast in a pandas world. Do you know how long it takes to run at the moment? You can say how long the tests take and how much data was in the tests, to get an eastimate on the year. I would say if its longer than 1 second  for a year of data then we should rewrite

Reply:
For a day of `5-minute` intervals, it is taking `0.17`sec, so for a year it must be like `62.05` sec, almost takes a minute for a year, I'm assuming that's bad!
![image](https://user-images.githubusercontent.com/93340339/208932516-b4770049-db46-4f71-815e-1873daf9579a.png)

Reply:
thanks for running that, yea, hmmm, bit slow, maybe worth extning the test to a week, and see how long it takes (and then time sby 52). There may be some over heads with setupping the test

Reply:
I have added `numba` `@jit decorator` and `numba.prange` for the `for loop`, as I read from the docs that numba makes the python code faster, and also added test to check the execution time with `timeit` module

Reply:
How long did it take with the extended time?

Reply:
Using `numba` and `@jit decorator`, the data types inside the `functional datapipe` are getting changed into `numba` datatypes, so when trying to use a combination of `numba` used `functional datapipe` and a regular `functional datapipe`, it is showing the error as not compatible data types. I have checked the execution time for both cases `numba` and regular functions, does not have much difference, atm with the current code it is taking 13. seconds for 365 iterations.